### PR TITLE
allow connecting to multiple LDAP services

### PR DESCRIPTION
**Pull request by unki**
Grafana allows specifying [multiple LDAP connections](https://grafana.com/docs/grafana/latest/auth/ldap/#multiple-ldap-servers) within _ldap.toml_.

This is a bit tricky within a hash-structure - Puppet code as well as hiera/yaml - as the keys for `servers`, `servers.attributes`, etc. would then repeat.

Also it is then required, that the sections for `servers.attributes` and `servers.group_mappings` are tied to the `servers` section.

I've accomplished it now by enwrapping the configurations with a surrounding array - therefor accepting hashes and arrays now for `$ldap_cfg`. Of course using a hash to enwrap would also be possible.

Is this a way to go?

### CODE REVIEW

**Comment by bastelfreak** [comment]
Hi @unki, thanks for the PR. Can you please add some tests for this?

**Comment by unki** [comment]
> Hi @unki, thanks for the PR. Can you please add some tests for this?

Hi @bastelfreak, thanks for considering the PR!
Finally I got the test run successful - please take a look if they are ok for you.

File: manifests/init.pp [hunk 1]
```diff
@@ -110,7 +110,7 @@
   Optional[String] $archive_source,
   String $cfg_location,
   Hash $cfg,
-  Optional[Hash] $ldap_cfg,
+  Optional[Variant[Hash,Array]] $ldap_cfg,
```
Comment (Dan33l) [inline manifests/init.pp]: It would be great if we build a custom data type here.

Comment (bastelfreak) [reply manifests/init.pp]: I would also be fine with:
```suggestion
  Optional[Variant[Hash[String[1],String[1]],Array][String[1]]] $ldap_cfg,
```

(assuming that the content is always strings)

**Comment by unki** [comment]
@bastelfreak @Dan33l actually I'm was already working on custom-types now and almost done with creating a kind of a prototype (haven't written tests yet).

Currently corresponding puppet-code would then look like below - borrowed a bit how _puppetlabs-concat_ and _puppetlabs-mysql_ are handling comparable stuff:

```puppet

# that one seems a bit superfluous, put I was needing one type where I can collect
# all the child-resources and create the toml out of
grafana_ldap_config { '/etc/grafana/config.toml': }

grafana_ldap_server {
  'foobar':
    hosts                              => [
      'host1.example.com',
      'host1a.example.com'
    ],
    use_ssl                            => true,
    start_tls                          => false,
    ssl_skip_verify                    => true,
    root_ca_cert                       => '/etc/foo',
    client_cert                        => '/usr/bar',
    client_key                         => '/tmp/key',
    bind_dn                            => 'username',
    bind_password                      => 'password',
    user_search_filter                 => 'uid=%s',
    user_search_base_dns               => ['ou=users,dc=example,dc=com'],
    group_search_filter                => 'objectClass=posixGroup',
    group_search_filter_user_attribute => 'memberUid=%s',
    group_search_base_dns              => ['ou=groups,dc=example,dc=com'],
    attributes                         => {
      name      => 'givenName',
      surname   => 'sn',
      username  => 'cn',
      member_of => 'memberOf',
      email     => 'email',
    }, ;

  'blabla':
    hosts                => ['host2.example.com'],
    user_search_base_dns => ['ou=users,dc=example2,dc=com'], ;
}

grafana_ldap_group_mapping {
  'foobar_cn=admin':
    ldap_server_config => 'foobar',
    group_dn           => 'cn=admin',
    org_role           => 'Admin',
    grafana_admin      => true, ;
  'foobar_cn=editor':
    ldap_server_config => 'foobar',
    group_dn           => 'cn=editor',
    org_role           => 'Editor', ;
  'foobar_cn=viewer':
    ldap_server_config => 'foobar',
    group_dn           => 'cn=viewer',
    org_role           => 'Viewer', ;

  'blabla_cn=viewer':
    ldap_server_config => 'blabla',
    group_dn           => 'cn=viewer',
    org_role           => 'Viewer', ;
}
```

Would that be ok to continue?

**Comment by bastelfreak** [comment]
I'm always :+1: for custom types!

File: manifests/install.pp [hunk 2]
```diff
@@ -195,4 +195,9 @@
       fail("Installation method ${grafana::install_method} not supported")
     }
   }
+
+  if !empty($grafana::ldap_servers) {
```
Comment (bastelfreak) [inline manifests/install.pp]: I'm not sure if this whole if block is a good idea. the ruby-toml package isn't really grafana specific, so people might already declare it in another module. What about a new parameter `Boolean $manage_toml` that wraps `ensure_package()` from stdlib?

Comment (unki) [reply manifests/install.pp]: Sure, no problem. Have had discovered in the morning that I need ruby-toml, as the generating part in the custom-type moves then from puppetserver to agents.

But I was also thinking if it is actually worth the effort to add ruby-toml as dependency on agent-side - or using an ordinary puppet-template instead to generate the config. Not sure.

File: manifests/init.pp (hunk 1 above)
Comment (bastelfreak) [reply manifests/init.pp]: ping @unki :)

File: manifests/init.pp [hunk 3]
```diff
@@ -147,4 +180,7 @@
   # Dependency added for Grafana_plugins to ensure it runs at the
   # correct time.
   Class['grafana::config'] -> Grafana_Plugin <| |> ~> Class['grafana::service']
+
+  create_resources('grafana_ldap_server', $ldap_servers)
```
Comment (ghoneycutt) [inline manifests/init.pp:184]: If you're interested, here's a more modern approach to `create_resources()`. No need to change this though.

https://github.com/sensu/sensu-puppet/blob/master/manifests/resources.pp#L78-L82

**Pull request merged**